pkg/agent/reconcilers.go

```diff
@@ -124,65 +123,51 @@ func (a *Agent) handleAppliedManifestWork(obj runtime.Object, isBeingDeleted boo
+func (a *Agent) updateWorkStatus(obj runtime.Object) error {
 	mObj := obj.(metav1.Object)
 	namespace := a.clusterName
 
-	workStatus := &v1alpha1.WorkStatus{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      util.BuildWorkstatusName(obj),
-			Namespace: namespace,
-		},
-	}
-
 	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
 	defer cancel()
 
-	// delete WorkStatus if exists, when the workload object is deleted
-	if isBeingDeleted {
-		err := a.hubClient.Get(ctx, client.ObjectKeyFromObject(workStatus), workStatus, &client.GetOptions{})
-		if err != nil && !apierrors.IsNotFound(err) {
-			return err
-		}
-		err = a.hubClient.Delete(ctx, workStatus, &client.DeleteOptions{})
-		if err != nil {
-			a.logger.Info("workStatus was previously deleted", "workStatus-name", workStatus.Name)
-			return nil
-		}
-		a.logger.Info("workStatus deleted", "workStatus-name", workStatus.Name)
-		return nil
+	aWork, err := ocm.GetAppliedManifestWork(obj, a.listers)
+	if err != nil || aWork == nil {
+		return fmt.Errorf("AppliedManifestWork not found for object with name=%s", mObj.GetName())
 	}
 
-	name, ok := a.trackedObjects.Get(string(mObj.GetUID()))
-	if !ok {
-		return fmt.Errorf("object not found in tracked objects: uid=%s", string(mObj.GetUID()))
+	// init workstatus object
+	workStatus := &v1alpha1.WorkStatus{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      util.BuildWorkstatusName(*aWork, obj),
+		},
 	}
 
 	// check if WorkStatus exists and if not create it
-	err := a.hubClient.Get(ctx, client.ObjectKeyFromObject(workStatus), workStatus, &client.GetOptions{})
+	err = a.hubClient.Get(ctx, client.ObjectKeyFromObject(workStatus), workStatus, &client.GetOptions{})
 	if err != nil {
 		if apierrors.IsNotFound(err) {
 			// get the manifest work for this workstatus, so that we can set a owner ref
-			manifestWork, err := ocm.GetManifestWork(a.hubClient, name, namespace)
+			manifestWork, err := ocm.GetManifestWork(a.hubClient, aWork.Spec.ManifestWorkName, namespace)
 			if err != nil {
 				return fmt.Errorf("failed to get manifestWork: %w", err)
 			}
 
 			// only update status for KS-managed (by bindingpolicies here) objects
 			// TODO remove the check for ManagedByKSLabelKeyPrefix after we switch to new transport
-			if !util.HasPrefixInMap(manifestWork.Labels, ManagedByKSLabelKeyPrefix) && !util.HasPrefixInMap(manifestWork.Labels, TransportLabelPrefix) {
-				a.logger.Info("object not managed by a KS bindingpolicy, status not updated", "object", name, "namespace", namespace)
+			if !(util.HasPrefixInMap(manifestWork.Labels, ManagedByKSLabelKeyPrefix) || util.HasPrefixInMap(manifestWork.Labels, TransportLabelPrefix)) {
+				a.logger.Info("object not managed by a KS bindingpolicy, status not updated", "object", aWork.Spec.ManifestWorkName, "namespace", namespace)
 				return nil
 			}
 
+			// this was causing race conditions
```

Comment:
Maybe this comment should be removed if this is no longer causing race conditions.

Reply:
done